analyzers/src/SonarAnalyzer.CSharp/Rules/CollectionsShouldImplementGenericInterface.cs

```diff
@@ -34,25 +34,28 @@ namespace SonarAnalyzer.Rules.CSharp
     [DiagnosticAnalyzer(LanguageNames.CSharp)]
     public sealed class CollectionsShouldImplementGenericInterface : SonarDiagnosticAnalyzer
     {
-        internal const string DiagnosticId = "S3909";
+        private const string DiagnosticId = "S3909";
         private const string MessageFormat = "Refactor this collection to implement '{0}'.";
 
         private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorBuilder.GetDescriptor(DiagnosticId, MessageFormat, RspecStrings.ResourceManager);
 
-        private static readonly Dictionary<string, KnownType> NongenericToGenericMapping = new()
+        private static readonly Dictionary<KnownType, string> NonGenericToGenericMapping = new()
         {
-            { KnownType.System_Collections_ICollection.TypeName, KnownType.System_Collections_Generic_ICollection_T },
-            { KnownType.System_Collections_IList.TypeName, KnownType.System_Collections_Generic_IList_T },
-            { KnownType.System_Collections_IEnumerable.TypeName, KnownType.System_Collections_Generic_IEnumerable_T },
-            { KnownType.System_Collections_CollectionBase.TypeName, KnownType.System_Collections_ObjectModel_Collection_T },
+            { KnownType.System_Collections_ICollection, "System.Collections.Generic.ICollection<T>" },
+            { KnownType.System_Collections_IList, "System.Collections.Generic.IList<T>" },
+            { KnownType.System_Collections_IEnumerable, "System.Collections.Generic.IEnumerable<T>" },
+            { KnownType.System_Collections_CollectionBase, "System.Collections.ObjectModel.Collection<T>" },
         };
 
-        private static readonly ImmutableArray<KnownType> GenericTypes = NongenericToGenericMapping.Values.ToImmutableArray();
+        private static readonly ImmutableArray<KnownType> GenericTypes = ImmutableArray.Create(
+            KnownType.System_Collections_Generic_ICollection_T,
+            KnownType.System_Collections_Generic_IList_T,
+            KnownType.System_Collections_Generic_IEnumerable_T,
+            KnownType.System_Collections_ObjectModel_Collection_T);
```

Comment:
They are different (generic)

Reply:
Ah, ok. And can we make the `NonGenericToGenericMapping` a dictionary of `<KnownType, KnownType>` instead of `<KnownType, string>`? Like that, we could remove this field completely.

Becauase AFAICT, the dictionary is not really needed there. It's just a collection of pairs (tuples).

Maybe in another PR.

Reply:
Here, we could benefit from the `.ToString()` or `.FullName` behavior

Reply:
Done in https://github.com/SonarSource/sonar-dotnet/pull/5699

Reply:
We've decided not to do this change. The generic vs C# vs VB introduces too much complexity.